DOCS-544: Program Robots section refactor (#1364)
viamrobotics/docs: docs/program/_index.md

linkTitle: "Program Robots"
childTitleEndOverwrite: "Program Robots"
description: "Use the SDK of your preferred language to write code to control your robots."
weight: 45
no_list: true
type: docs
images: ["/img/code.png"]
aliases:
  - "product-overviews/sdk-as-client"
  - "program/sdk-as-client"
  - "program/sdks"
---

Viam offers software development kits (SDKs) in popular languages which

> - Streamline connection, authentication, and encryption against a server using {{< glossary_tooltip term_id="webrtc" >}}

Author: I like that phrasing! Edited to add code links.

> - Enable you to interface with robots without calling the `viam-server` [gRPC APIs for robot controls](https://github.com/viamrobotics/api) directly

Author: @mcvella Ah ok if you think that's better terminology it's a lot cleaner, so done

![Diagram showing how a client connects to a robot with Viam. Diagram shows a client as a computer sending commands to a robot. Robot 1 then communicates with other robotic parts over gRPC and WebRTC and communicating that information back to the client.](img/sdks/robot-client.png)

Use the SDK of your preferred language to write code to control your robots.

> Viam currently offers SDKs for the following languages:

Author: I am not sure on this actually. @npentrel LMK if you see this and have a confirmation on this

> Viam currently offers SDKs for the following languages:

Author: Andrew is covering this in a separate PR today!

> Viam currently offers SDKs for the following languages:

Author: Done

- [Python SDK](https://python.viam.dev/)
- [Go SDK](https://pkg.go.dev/go.viam.com/rdk)
- [TypeScript SDK](https://ts.viam.dev/)
- [C++ SDK (alpha)](https://cpp.viam.dev/)

## Requirements

> Before you get started with your program, ensure that you have [installed and connected to `viam-server`](/installation/) on the computer you want to use to control a robot [(likely a single-board computer)](/components/board/#configuration) and [configured a robot](/manage/configuration/#local-setup).

Author: @mcvella Good catch here. I think that just had been up and I never questioned it, so I don't know why either!

> The easiest way to do this is:

Author: Good point, I like having those instructions there for beginners but I don't think it's necessary in this case with the other intro paragraph.

1. Go to [app.viam.com](https://app.viam.com/).
2. Create a new robot.
3. Go to the **Setup** tab and follow the instructions there.
4. [Configure](/manage/configuration/) your robot.

> Next, to install your preferred Viam SDK on your development machine, run one of the following commands in your terminal:

Reviewer: Do we want to here mention that said development machine must be Linux or Mac?
Also, they could run SDKs from their robots (SBCs) themselves, and there are many valid reasons to do so.  I think we should mention this as an option.

Author: Great points, done (I think)

{{< tabs >}}
{{% tab name="Python" %}}

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
pip install viam-sdk
```

{{% /tab %}}
{{% tab name="Go" %}}

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
go install go.viam.com/rdk/robot/client@latest
```

{{% /tab %}}
{{% tab name="TypeScript" %}}

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
npm install --save @viamrobotics/sdk
```

{{% /tab %}}
{{% tab name="C++" %}}

Follow the [instructions on the GitHub repository](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/BUILDING.md).

{{% /tab %}}
{{< /tabs >}}

## Hello World: The Code Sample Tab

> Now, create a program in the language of your choice that connects to your robot and uses methods built into the SDK's client API libraries to [interact with and control](/program/sdks/) the {{< glossary_tooltip term_id="resource" text="resources" >}} on the robot.

Reviewer 2: ```suggestion
Create a program in the language of your choice that connects to your robot and uses methods built into the SDK's client API libraries to [interact with and control](/program/sdks/) the {{< glossary_tooltip term_id="resource" text="resources" >}} on the robot.
```

(They may be linked to this section and start here)

Author: Good point, addressed

Start with this by navigating to your robot's page on [the Viam app](https://app.viam.com/robots).
Select the **Code Sample** tab, select your preferred SDK, and copy the code generated for you.

> These code snippets import all the necessary libraries and set up a connection with your robot in the cloud through the Viam app.

Reviewer: ```suggestion
This code snippet imports all the necessary libraries to set up a connection with your robot and interface with its configured components and services.
```
Since your robot is not usually in the cloud and the generated code does create methods for all viam-server components and services (not modular ones at the moment)

    return;
  }
}

main();
```

{{% /tab %}}
{{% tab name="C++" %}}

{{< alert title="Note" color="note" >}}
The C++ SDK is currently in alpha.
{{< /alert >}}

```cpp {class="line-numbers linkable-line-numbers"}
# include <string>
# include <vector>

# include <boost/optional.hpp>

# include <viam/api/common/v1/common.pb.h>
# include <viam/api/robot/v1/robot.grpc.pb.h>

# include <viam/sdk/robot/client.hpp>
# include <viam/sdk/components/camera/client.hpp>

using namespace viam::sdk;

int main() {
  std::string host("ADDRESS FROM THE VIAM APP");
  DialOptions dial_opts;
  Credentials credentials("SECRET FROM THE VIAM APP");
  dial_opts.set_credentials(credentials);
  boost::optional<DialOptions> opts(dial_opts);
  Options options(0, opts);

  auto robot = RobotClient::at_address(host, options);

  std::cout << "Resources:\n";
  for (const ResourceName& resource: *robot->resource_names()) {
    std::cout << resource.namespace_() << ":" << resource.type() << ":"
              << resource.subtype() << ":" << resource.name() << "\n";
  }

  return 0;
}
```

{{% /tab %}}
{{< /tabs >}}
{{% /expand%}}

Save this file on your development machine with the file type of your preferred SDK.

> Edit the boilerplate code by [adding control logic](/program/sdks/) to make a client application that connects to your robot and controls it in the way you want.

Reviewer: I think this sentence actually should come later.  Typically one would run the sample code-as is first to ensure requirements are met and connectivity works, then they'd alter it to suit their needs

Author: @mcvella That is a good point re. why that would be needed. Done!

Then, [execute this program](/program/run/) on any computer which:

1. has [the appropriate SDK](#requirements) installed

> 2. can establish a connection to your robot through the cloud, on a local or wide area network (LAN or WAN), or [locally](/program/run/#run-code-locally)

Author: ```suggestion
2. can establish a connection to your robot through the cloud, on a local or wide area network (LAN or WAN), or [locally](/program/run/#run-code-locally)

Edit the boilerplate code by [adding control logic](/program/sdks/) to make a client application that connects to your robot and controls it in the way you want.
```

> Then, [execute this program](/program/run/) on any computer as long as:
> 
> 1. You have [installed the appropriate SDK](#requirements) and language on this machine.
> 2. The program establishes a connection to your robot through the cloud, on a local or wide area network (LAN or WAN), or [locally](/program/run/#run-code-locally) when [executed](/program/run/).

Reviewer 2: ```suggestion
Then, [execute this program](/program/run/) on any computer which:

- has [the appropriate SDK](#requirements) installed
- can establish a connection to your robot through the cloud, on a local or wide area network (LAN or WAN), or [locally](/program/run/#run-code-locally)
```

Author: Done

### `import`: Find Required Imports

> The Code Sample tab contains the required imports for the SDK code at generation.

Reviewer: What is this meant to mean?

> ### `import`: Find Required Imports

Reviewer 2: I think this and the following section could instead be embedded in the text before you say "Edit the boilerplate" because the you explain the boilerplate first. If you want to add section headings maybe "Copy the Code Sample" and "Edit and run your code" would work better?

Reviewer 2: thoughts?

Reviewer: I think editing the boilerplate should come in this later section, the above sections should strictly be setting up and running the boilerplate as it was generated.

Author: @mcvella @npentrel Is this ok as is then? I don't have a strong opinion either way!

Reviewer 2: I think Matt and I both think the ordering is weird (and I do also think the headlines are confusing - because you're not finding imports, they should already be there). How about this instead:

```
## Hello World: The Code Sample Tab

Create a program in the language of your choice that connects to your robot and uses methods built into the SDK's client API libraries to [interact with and control](/program/sdks/) the {{< glossary_tooltip term_id="resource" text="resources" >}} on the robot.

Start by navigating to your robot's page on [the Viam app](https://app.viam.com/robots).
Select the **Code Sample** tab, select your preferred SDK, and copy the code generated for you.
This code snippet imports all the necessary libraries to set up a connection with your robot and interface with its configured components and services.

Your boilerplate code sample should look similar to this:

{{%expand "Click this to see example boilerplate code from the Code Sample tab" %}}
...
{{% /expand%}}

Save this file on your development machine with the file type of your preferred SDK.

The sample code contains the required imports as well as the connect logic which establishes a connection for your client application to [communicate with](/internals/robot-to-robot-comms/) the robot's `viam-server` instance.
This section of the boilerplate code contains your robot's address and location secret.
You can think of these as keys or access tokens to your robot that are important to keep private.
This connection must be established for your program to be executed properly on your robot.

{{< readfile "/static/include/snippet/secret-share.md" >}}

### Run the sample code

Once you have saved the sample code, [execute your program](/program/run/).

You can run your program on any computer which:

1. has [the appropriate SDK](#requirements) installed
2. can establish a connection to your robot through the cloud, on a local or wide area network (LAN or WAN), or [locally](/program/run/#run-code-locally)

The program will connect to your robot and print a list of the available {{< glossary_tooltip term_id="resources" text="resources" >}}.

### Edit the sample code

Once you have successfully run the sample code, you can edit the boilerplate code by [adding control logic](/program/sdks/) to make a client application that connects to your robot and controls it in the way you want.
You can find the right libraries to import for SDK methods, typing, interfaces, and utilities at the start of [each resource's API documentation](/program/sdks/resource-apis/), as well as in the individual SDK documentation sites and [on GitHub](https://github.com/viamrobotics/rdk).
```

Author: Done, thanks for taking the time to type that out!

> ### `import`: Finding Required Imports

Author: ```suggestion
### `import`: Find Required Imports
```

If you are building out your program further or aren't using the Code Sample tab, you can find the right libraries to import to utilize SDK methods, typing, interfaces, and utilities at the start of [each resource's API documentation](/program/sdks/resource-apis/), as well as in the individual SDK documentation sites and [on GitHub](https://github.com/viamrobotics/rdk).

> ### `connect`: Connecting with Your Robot

Author: ```suggestion
### `connect`: Connect with Your Robot
```

The `connect` logic of an SDK program establishes a connection for your client application to [communicate with](/internals/robot-to-robot-comms/) the robot's `viam-server` instance.

> This section of the boilerplate code contains your robot's address, location secret, and robot secret.

Reviewer: Isn't the location and robot secret synonymous?

Author: So basically it was inaccurate to include in this sentence and should have been `robot part secret` but they are two separate things as far as I can tell-- see Security tab in app for the part secrets, and code sample tab for the location secret.

linkTitle: "Program Robots"
childTitleEndOverwrite: "Program Robots"
description: "Use the SDK of your preferred language to write code to control your robots."
weight: 45
no_list: true
type: docs
images: ["/img/code.png"]
aliases:
  - "product-overviews/sdk-as-client"
  - "program/sdk-as-client"
  - "program/sdks"
---

Viam offers software development kits (SDKs) in popular languages which

- Brokers connection, authentication, and encryption for communication with robots running `viam-server` using {{< glossary_tooltip term_id="webrtc" >}}

> - Enables you to interface with robot [gRPC APIs](https://github.com/viamrobotics/api) in a way that is idiomatic to that programming language

Author: No worries, done

  DialOptions dial_opts;
  Credentials credentials("SECRET FROM THE VIAM APP");
  dial_opts.set_credentials(credentials);
  boost::optional<DialOptions> opts(dial_opts);
  Options options(0, opts);

  auto robot = RobotClient::at_address(host, options);

  std::cout << "Resources:\n";
  for (const ResourceName& resource: *robot->resource_names()) {
    std::cout << resource.namespace_() << ":" << resource.type() << ":"
              << resource.subtype() << ":" << resource.name() << "\n";
  }

  return 0;
}
```

{{% /tab %}}
{{< /tabs >}}
{{% /expand%}}

Save this file on your development machine with the file type of your preferred SDK.

Edit the boilerplate code by [adding control logic](/program/write/) to make a client application that connects to your robot controls it in the way you want when the program [is executed](/program/run/).

Then, execute this program on any computer as long as:

1. You have [installed the appropriate SDK](#requirements) and language on this machine.
2. The program establishes a connection to your robot through the cloud, on a local or wide area network (LAN or WAN), or [locally](/program/run/#run-code-locally) when [executed](/program/run).

> See [Run Code](/run) for more information on executing code.

Author: ```suggestion
See [Run Code](/program/run) for more information on executing code.
```

### `import`: Finding Required Imports

The Code Sample tab contains the required imports for the SDK code at generation.

> If you are building out your program further or aren't using the Code Sample tab, you can find the right libraries to import to utilize SDK methods, typing, interfaces, and utilities at the start of [each resource's API documentation](/program/write/resource-apis/), as well as in the individual SDK documentation sites and [on GitHub](https://github.com/viamrobotics/rdk).

Author: ```suggestion
If you are building out your program further or aren't using the Code Sample tab, you can find the right libraries to import to utilize SDK methods, typing, interfaces, and utilities at the start of [each resource's API documentation](/program/sdks/resource-apis/), as well as in the individual SDK documentation sites and [on GitHub](https://github.com/viamrobotics/rdk).
```

### `connect`: Connecting with Your Robot

The `connect` logic of an SDK program establishes a connection for your client application to [communicate with](/internals/robot-to-robot-comms/) the robot's `viam-server` instance.
This section of the boilerplate code contains your robot's address, location secret, and robot secret.

> You can think of these as key or access tokens to your robot that are important to keep private.

Author: ```suggestion
You can think of these as keys or access tokens to your robot that are important to keep private.
```

// Access a component
final movementSensor = MovementSensor.fromRobot(robot, 'my_sensor');
print(await movementSensor.readings())
```

{{% /tab %}}
{{< /tabs >}}
{{% /expand%}}

Save this file on your development machine with the file type of your preferred SDK.

The sample code contains the required imports as well as the connect logic which establishes a connection for your client application to [communicate with](/internals/robot-to-robot-comms/) the robot's `viam-server` instance.
This section of the boilerplate code contains your robot's address and location secret.
You can think of these as keys or access tokens to your robot that are important to keep private.
This connection must be established for your program to be executed properly on your robot.

{{< readfile "/static/include/snippet/secret-share.md" >}}

### Run the sample code

Once you have saved the sample code, [execute your program](/program/run/).

You can run your program on any computer which:

1. has [the appropriate SDK](#requirements) installed
2. can establish a connection to your robot through the cloud, on a local or wide area network (LAN or WAN), or [locally](/program/run/#run-code-locally)

> The program will connect to your robot and print a list of the available {{< glossary_tooltip term_id="resources" text="resources" >}}.

Author: ```suggestion
The program will connect to your robot and print a list of the available {{< glossary_tooltip term_id="resource" text="resources" >}}.
```

Viam offers software development kits (SDKs) in popular languages which

- Brokers connection, authentication, and encryption for communication with robots running `viam-server` using {{< glossary_tooltip term_id="webrtc" >}}

> - Enable you to interface with robot [gRPC APIs](https://github.com/viamrobotics/api) in a way that is idiomatic to that programming language

Reviewer 2: ```suggestion
- Enables you to interface with robot [gRPC APIs](https://github.com/viamrobotics/api) in a way that is idiomatic to that programming language
```

## Requirements

> Before you get started with your program, ensure that you have [installed `viam-server`](/installation/) on the computer you want to use to control a robot [(likely a single-board computer)](/components/board/#configuration) and [configured a robot](/manage/configuration/).

Reviewer 2: ```suggestion
Before you get started with your program, ensure that you have [installed `viam-server`](/installation/) on the computer you want to use to control your robot [(likely a single-board computer)](/components/board/#configuration) and [configured a robot](/manage/configuration/).
```
